## PешАлово 2.0 или Что нужно государству?


Author:  olga.zagovora@gesis.org
    
Data from Serhiy Protsenko here https://github.com/SerhiyProtsenko/Magister_2018_K1-K9 

In [1]:
from collections import defaultdict
import pandas as pd
import os
import fnmatch
import json
import sys
stdout = sys.stdout
reload(sys)
sys.setdefaultencoding('utf-8')
sys.stdout = stdout

In [2]:
df=pd.read_excel("Mag2018_K1_k9_v2.xlsx")
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3013 entries, 1307 to 38
Data columns (total 23 columns):
рік                                                         3013 non-null int64
код ЗВО                                                     3013 non-null int64
Назва ЗВО                                                   3013 non-null object
Код спец.                                                   3013 non-null int64
Код. спеціал                                                3013 non-null int64
Назва спец.                                                 3013 non-null object
K1                                                          3013 non-null float64
K2                                                          3013 non-null float64
K3                                                          3013 non-null float64
K4                                                          3013 non-null float64
K5                                                          3013 non-null float64


In [3]:
import statsmodels.formula.api as smf
import statsmodels.api as sm

In [4]:
just_dummies = pd.get_dummies(df[u'Назва ЗВО'])#,drop_first=True)
df_1 = pd.concat([df, just_dummies], axis=1)      
#df_1.drop(['Назва ЗВО'], inplace=True, axis=1)
df_1.rename(columns={u'Обсяг державного замовлення Всього, денна':"ODZ",u'Mult K1-K9':"Mult_K1_K9",
                     u'Випуск бак., денна':"VB",u'Випуск бак. min 5, денна':"VB5",
                     u'Назва ЗВО':"Nazva_ZVO",u'Назва спец.':"Nazva_spec"}, inplace=True)
df_2=df_1[[u'Mult_K1_K9',"VB","ODZ"]]
df_2=sm.add_constant(df_2)

In [5]:
y=df_1["ODZ"]

In [6]:
result = smf.ols(formula='ODZ ~ Mult_K1_K9 + VB', data=df_1).fit()
#result = sm.OLS(y,df_1[["const","Mult_K1_K9" ,"VB"]]).fit()
print result.summary()

                            OLS Regression Results                            
Dep. Variable:                    ODZ   R-squared:                       0.877
Model:                            OLS   Adj. R-squared:                  0.877
Method:                 Least Squares   F-statistic:                 1.071e+04
Date:                Mon, 23 Jul 2018   Prob (F-statistic):               0.00
Time:                        17:21:19   Log-Likelihood:                -10096.
No. Observations:                3013   AIC:                         2.020e+04
Df Residuals:                    3010   BIC:                         2.022e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -7.6594      0.968     -7.909      0.0

In [7]:
df_1[["Nazva_ZVO",u'Mult_K1_K9',u'ODZ',"VB"]].head()

Nazva_ZVO  Mult_K1_K9  ODZ  VB
1307  Донецький нацiональний унiверситет ім. Василя ...    2.212647    6   3
2060  Київський національний університет ім. Тараса ...    2.075149   36  56
2418  Харківський національний унiверситет iм. В.Н.К...    1.960006   16  16
634                      Сумський державний унiверситет    1.872965   11  11
1569  Львівський нацiональний унiверситет імені Іван...    1.855975   21  23

In [8]:
6.7655*1.87+0.5180*11-7.6594 #R^2 0.877

10.690085000000002

In [9]:
result = smf.ols(formula='ODZ ~ Mult_K1_K9 + VB5', data=df_1).fit()
print result.summary()

                            OLS Regression Results                            
Dep. Variable:                    ODZ   R-squared:                       0.881
Model:                            OLS   Adj. R-squared:                  0.881
Method:                 Least Squares   F-statistic:                 1.111e+04
Date:                Mon, 23 Jul 2018   Prob (F-statistic):               0.00
Time:                        17:22:27   Log-Likelihood:                -10047.
No. Observations:                3013   AIC:                         2.010e+04
Df Residuals:                    3010   BIC:                         2.012e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -8.5319      0.952     -8.958      0.0

In [11]:
6.9379*1.87+0.5265*11-8.5319 #R^2 0.881

10.233473

In [12]:
result = smf.ols(formula='ODZ ~ Mult_K1_K9*VB', data=df_1).fit()
print result.summary()

                            OLS Regression Results                            
Dep. Variable:                    ODZ   R-squared:                       0.893
Model:                            OLS   Adj. R-squared:                  0.892
Method:                 Least Squares   F-statistic:                     8333.
Date:                Mon, 23 Jul 2018   Prob (F-statistic):               0.00
Time:                        17:23:18   Log-Likelihood:                -9888.9
No. Observations:                3013   AIC:                         1.979e+04
Df Residuals:                    3009   BIC:                         1.981e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept         7.2324      1.149      6.297

In [13]:
0.4033*1.87*11 - 0.1219*11  -2.9160*1.87 + 7.2324    #R^2 0.893

8.734461

In [14]:
result = smf.ols(formula='ODZ ~ Mult_K1_K9*VB5', data=df_1).fit()
print result.summary()

                            OLS Regression Results                            
Dep. Variable:                    ODZ   R-squared:                       0.895
Model:                            OLS   Adj. R-squared:                  0.895
Method:                 Least Squares   F-statistic:                     8540.
Date:                Mon, 23 Jul 2018   Prob (F-statistic):               0.00
Time:                        17:24:30   Log-Likelihood:                -9855.9
No. Observations:                3013   AIC:                         1.972e+04
Df Residuals:                    3009   BIC:                         1.974e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept          6.3877      1.160      5.

In [15]:
df_1["Nazva_ZVO"]=df_1.Nazva_ZVO.apply(lambda x: x.split()[0]+" "+"".join(item[0].upper() for item in x.split()[1:]))


from transliterate import translit
df_1["Nazva_ZVO"]=df_1["Nazva_ZVO"].apply(lambda x: translit(x, "uk", reversed=True))

In [16]:
result = smf.ols(formula='ODZ ~ Mult_K1_K9*VB5 + C(Nazva_ZVO)', data=df_1).fit()
print result.summary()

                            OLS Regression Results                            
Dep. Variable:                    ODZ   R-squared:                       0.911
Model:                            OLS   Adj. R-squared:                  0.907
Method:                 Least Squares   F-statistic:                     206.5
Date:                Mon, 23 Jul 2018   Prob (F-statistic):               0.00
Time:                        17:24:49   Log-Likelihood:                -9597.8
No. Observations:                3013   AIC:                         1.948e+04
Df Residuals:                    2869   BIC:                         2.035e+04
Df Model:                         143                                         
Covariance Type:            nonrobust                                         
                                                      coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------

In [17]:
#print 0.3430*1.87 * 11  + 6.2199
#print 0.3430*1.87 * 11 - 2.3647*1.87 -0.0284*11  + 6.2199

sumDU_k1k9_mev=df_1[df_1[u'Nazva_ZVO']=="Sums'kyj DU"][u'Mult_K1_K9'].iloc[0]
sumDU_bak_mev=df_1[df_1[u'Nazva_ZVO']=="Sums'kyj DU"][u'VB5'].iloc[0]

print 0.3477*sumDU_k1k9_mev * sumDU_bak_mev -3.4599*sumDU_k1k9_mev  + 7.6674
print 0.3477*sumDU_k1k9_mev * sumDU_bak_mev -3.4599*sumDU_k1k9_mev - 0.0229*sumDU_bak_mev  + 7.6674

8.35065756196
8.09875756196


In [18]:
df[df[u'Назва ЗВО']=="Сумський державний унiверситет"]#[u'Mult K1-K9']#.iloc[0]

рік  код ЗВО                       Назва ЗВО  Код спец.  Код. спеціал  \
634  2018      168  Сумський державний унiверситет        292         11111   
633  2018      168  Сумський державний унiверситет        281         11111   
632  2018      168  Сумський державний унiверситет        171         11111   
631  2018      168  Сумський державний унiверситет        163         11111   
630  2018      168  Сумський державний унiверситет        161         11111   
629  2018      168  Сумський державний унiверситет        153         11111   
628  2018      168  Сумський державний унiверситет        152         11111   
627  2018      168  Сумський державний унiверситет        151         11111   
626  2018      168  Сумський державний унiверситет        144         11111   
625  2018      168  Сумський державний унiверситет        142         11111   
624  2018      168  Сумський державний унiверситет        141         11111   
623  2018      168  Сумський державний унiверситет        133         11111   
622  2018      168  Сумський державний унiверситет        132         11111   
621  2018      168  Сумський державний унiверситет        131         11111   
620  2018      168  Сумський державний унiверситет        122         11111   
619  2018      168  Сумський державний унiверситет        113         11111   
618  2018      168  Сумський державний унiверситет        101         11111   
617  2018      168  Сумський державний унiверситет         76         11111   
616  2018      168  Сумський державний унiверситет         75         11111   
615  2018      168  Сумський державний унiверситет         73         11111   
614  2018      168  Сумський державний унiверситет         72         11111   
613  2018      168  Сумський державний унiверситет         71         11111   
612  2018      168  Сумський державний унiверситет         61         11111   
611  2018      168  Сумський державний унiверситет         51         11111   
610  2018      168  Сумський державний унiверситет         35            41   
609  2018      168  Сумський державний унiверситет         17         11111   

                                           Назва спец.    K1    K2   K3    K4  \
634                    Міжнародні економічні відносини  1.25  1.07  1.1  1.04   
633             Публічне управління та адміністрування  1.25  1.07  1.1  1.04   
632                                        Електроніка  1.25  1.07  1.1  1.04   
631                               Біомедична інженерія  1.25  1.07  1.1  1.04   
630                    Хімічні технології та інженерія  1.25  1.07  1.1  1.04   
629                     Мікро- та наносистемна техніка  1.25  1.07  1.1  1.04   
628    Метрологія та інформаційно-вимірювальна техніка  1.25  1.07  1.1  1.04   
627           Автоматизація та комп’ютерно-інтегровані  1.25  1.07  1.1  1.04   
626                                    Теплоенергетика  1.25  1.07  1.1  1.04   
625                        Енергетичне машинобудування  1.25  1.07  1.1  1.04   
624               Електроенергетика, електротехніка та  1.25  1.07  1.1  1.04   
623                           Галузеве машинобудування  1.25  1.07  1.1  1.04   
622                                  Матеріалознавство  1.25  1.07  1.1  1.04   
621                                 Прикладна механіка  1.25  1.07  1.1  1.04   
620                                  Комп’ютерні науки  1.25  1.07  1.1  1.04   
619                               Прикладна математика  1.25  1.07  1.1  1.04   
618                                           Екологія  1.25  1.07  1.1  1.04   
617     Підприємництво, торгівля та біржова діяльність  1.25  1.07  1.1  1.04   
616                                          Маркетинг  1.25  1.07  1.1  1.04   
615                                         Менеджмент  1.25  1.07  1.1  1.04   
614          Фінанси, банківська справа та страхування  1.25  1.07  1.1  1.04   
613                              Облік і оподаткування  1.25  1.07  1.1  1.04   
612      

In [19]:
df_1["Nazva_spec"]=df_1["Nazva_spec"].apply(lambda x: translit(x, "uk", reversed=True))

df_1["Nazva_spec"]=df_1.Nazva_spec.apply(lambda x: x.split()[0]+" "+"".join(item[:3].title() for item in x.split()[1:]))

In [20]:
result = smf.ols(formula='ODZ ~ Mult_K1_K9*VB5 + C(Nazva_ZVO)+ C(Nazva_spec)', data=df_1).fit()
print result.summary()

                            OLS Regression Results                            
Dep. Variable:                    ODZ   R-squared:                       0.934
Model:                            OLS   Adj. R-squared:                  0.926
Method:                 Least Squares   F-statistic:                     122.7
Date:                Mon, 23 Jul 2018   Prob (F-statistic):               0.00
Time:                        17:28:12   Log-Likelihood:                -9162.6
No. Observations:                3013   AIC:                         1.895e+04
Df Residuals:                    2702   BIC:                         2.082e+04
Df Model:                         310                                         
Covariance Type:            nonrobust                                         
                                                       coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------

In [21]:
print "SumDU, Міжнародні економічні відносини (C(Nazva_spec)[T.Mizhnarodni EkoVid])"
print 0.3309 *sumDU_k1k9_mev *sumDU_bak_mev -8.4028 * 1  +  11.6803

SumDU, Міжнародні економічні відносини (C(Nazva_spec)[T.Mizhnarodni EkoVid])
10.0949046046


# Кому дали більше місць? І кого дискримінували?

In [22]:
s=result.pvalues
s_param01=result.params[result.params.index.isin(s[s<0.01].index)]

In [ ]:
# p val < 0.01

# ПорешАли?

In [23]:
s_param01[s_param01>0]

Intercept                                11.680343
C(Nazva_ZVO)[T.L'otna AN]                22.856043
C(Nazva_ZVO)[T.Natsional'nyj TUU"IIS]     4.442528
C(Nazva_ZVO)[T.Tavrijs'kyj DAU]           6.926547
C(Nazva_spec)[T.Aviatsijnyj Tra]         10.437244
Mult_K1_K9:VB5                            0.330901
dtype: float64

### Льотна академiя НАУ

### Національний технiчний унiверситет України "Київський політехнічний інститут ім.Ігоря Сікорського"

### Таврійський державний агротехнологічний університет


#### Спеціальність: Авіаційний транспорт


In [ ]:
# p val < 0.05

In [24]:
s_param05=result.params[result.params.index.isin(s[s<0.05].index)]
s_param05[s_param05>0]

Intercept                                          11.680343
C(Nazva_ZVO)[T.Khersons'ka DMA]                    14.392684
C(Nazva_ZVO)[T.L'otna AN]                          22.856043
C(Nazva_ZVO)[T.L'vivs'ka NAM]                       7.626252
C(Nazva_ZVO)[T.Natsional'nyj TUU"IIS]               4.442528
C(Nazva_ZVO)[T.Perejaslav-Khmel'nyts'kyj DPUHS]     3.695389
C(Nazva_ZVO)[T.Tavrijs'kyj DAU]                     6.926547
C(Nazva_spec)[T.Arkhitektura TaMis]                 4.132326
C(Nazva_spec)[T.Aviatsijnyj Tra]                   10.437244
C(Nazva_spec)[T.Profesijna Osv(Do]                 12.978226
Mult_K1_K9:VB5                                      0.330901
dtype: float64

### Херсонська  державна морська академія

### Льотна академiя НАУ

### Львівська національна академiя мистецтв

### Національний технiчний унiверситет України "Київський політехнічний інститут ім.Ігоря Сікорського"

### Переяслав-Хмельницький державний педагогiчний унiверситет ім. Григорія Сковороди

### Таврійський державний агротехнологічний університет




#### Спеціальності: Архітектура та містобудування, Авіаційний транспорт, Професійна освіта документознавство)

## Дискримінували (p<0.01)

In [25]:
s_param01[s_param01<0]

C(Nazva_ZVO)[T.Kyyivs'kyj NTU]                     -16.890687
C(Nazva_ZVO)[T.Luhans'kyj NAU]                      -5.778936
C(Nazva_ZVO)[T.Natsional'nyj U"MA]                 -10.452461
C(Nazva_ZVO)[T.Odes'kyj DAU]                        -7.259569
C(Nazva_ZVO)[T.Pivdennoukrayins'kyj NPUI]           -5.026378
C(Nazva_ZVO)[T.Poltavs'kyj NPUI]                    -4.935391
C(Nazva_spec)[T.Ahronomija ]                        -8.733813
C(Nazva_spec)[T.Avionika ]                          -8.720967
C(Nazva_spec)[T.Avtomatyzatsija TaKom]              -9.701148
C(Nazva_spec)[T.Avtomobil'nyj Tra]                  -8.946392
C(Nazva_spec)[T.Biomedychna Inz]                   -10.465256
C(Nazva_spec)[T.Budivnytstvo TaTsyInz]              -4.780298
C(Nazva_spec)[T.Doshkil'na Osv]                     -7.535465
C(Nazva_spec)[T.Dyzajn ]                           -13.154537
C(Nazva_spec)[T.Ekolohija ]                        -10.978703
C(Nazva_spec)[T.Ekonomika ]                        -12.330872
C(Nazva_

Дискримінація вузів

C(Nazva_ZVO)[T.Kyyivs'kyj NTU]                     -16.890687

C(Nazva_ZVO)[T.Luhans'kyj NAU]                      -5.778936

C(Nazva_ZVO)[T.Natsional'nyj U"MA]                 -10.452461

C(Nazva_ZVO)[T.Odes'kyj DAU]                        -7.259569

C(Nazva_ZVO)[T.Pivdennoukrayins'kyj NPUI]           -5.026378

C(Nazva_ZVO)[T.Poltavs'kyj NPUI]                    -4.935391
             

            
### Київський нацiональний торговельно-економiчний унiверситет

### Луганський національний аграрний університет

###  Національний унiверситет "Києво-Могилянська академiя"

### Одеський державний аграрний університет

### Пiвденноукраїнський національний педагогiчний унiверситет ім.К.Д.Ушинського
             
### Полтавський національний педагогiчний унiверситет ім.В.Г.Короленка
             
             
             
             